grand_challenge_forge/partials/example-evaluation-method/evaluate.py.j2

```diff
@@ -44,37 +47,154 @@ from helpers import run_prediction_processing, tree
+
+    # Make sure to save the metrics
+    write_metrics(metrics=metrics)
+
+    {% if phase.evaluation_additional_outputs %}
+
+    # For now, let us make bogus outputs for the evaluation
+    {%- for socket in phase.evaluation_additional_outputs  %}
+    output_{{ socket.slug.replace("-", "_")}} =
+        {%- if socket | has_example_value %}
+            {%- if socket.example_value is string -%}
+                "{{ socket.example_value }}"
+            {%- else -%}
+                {{ socket.example_value }}
+            {%- endif %}
+        {%- elif socket | is_image %} numpy.eye(4, 2)
+        {%- elif socket | is_json %} {"content": "should match the required format"}
+        {%- elif socket | is_file %} "content: should match the required format"
+        {% endif %}
+    {%- endfor %}
+
+    # Save your output
+    {% for socket in phase.evaluation_additional_outputs  -%}
+    {% set py_slug = socket.slug.replace("-", "_") -%}
+    {% if socket | is_image -%}
+    write_array_as_image_file(
+        location=OUTPUT_DIRECTORY / "{{ socket.relative_path }}",
+        array=output_{{ py_slug }},
+    )
+    {% endif -%}
+    {% if socket | is_json -%}
+    write_json_file(
+        location=OUTPUT_DIRECTORY / "{{ socket.relative_path }}",
+        content=output_{{ py_slug }}
+    )
+    {% endif -%}
+    {% if socket | is_file -%}
+    write_file(
+        location=OUTPUT_DIRECTORY / "{{ socket.relative_path }}",
+        content=output_{{ py_slug }}
+    )
+    {% endif -%}
+    {% endfor %}
+
+    {% endif %}
+
+    return 0
 
 def process(job):
+    # The key is a tuple of the slugs of the input sockets
+    interface_key = get_interface_key(job)
+
+    # Lookup the handler for this particular set of sockets (i.e. the interface)
+    handler = {
+    {% for interface_name, interface_key in algorithm_interface_names|zip(algorithm_interface_keys) -%}
+    {{ interface_key }} : process_{{ interface_name }},
+    {% endfor %}
+    }[interface_key]
+
+    {% if phase.evaluation_additional_inputs %}
+    # Read additional inputs from the submission
+
+    {% for socket in phase.evaluation_additional_inputs -%}
+    {% set py_slug = socket.slug.replace("-", "_") -%}
+    {% if socket | is_image -%}
+    submission_{{ py_slug }} = load_image_file_as_array(
+        location=INPUT_DIRECTORY / "{{ socket.relative_path }}",
+    )
+    {% endif -%}
+    {% if socket | is_json -%}
+    submission_{{ py_slug }} = load_json_file(
+        location=INPUT_DIRECTORY / "{{ socket.relative_path }}",
+    )
+    {% endif -%}
+    {% if socket | is_file -%}
+    submission_{{ py_slug }} = load_file(
+        location=INPUT_DIRECTORY / "{{ socket.relative_path }}",
+    )
+    {% endif -%}
+    {%- endfor %}
+    # Finally, call the handler
+    return handler(
+        job,
+    {% for socket in phase.evaluation_additional_inputs -%}
+       submission_{{ socket.slug.replace("-", "_") }},
+    {% endfor %}
+    )
+    {% else %}
+    # Call the handler
```

Comment:
I think this comment is duplicate (see line 149)

Reply:
I get your confusion. Normally, in Django templates, we'd indent the body of if/else. However, this is quite hard to do with Python code since it messes up the code.

Line 149 falls in the if-body (when there are evaluation inputs) and line 157 falls in the else-body (no evaluation inputs).

Reply:
Not sure if there is a better/cleaner way to do identation. I've found that adding a black call at the end of rendering already helps with the chaos: you just need valid Python code,  instead of having 'nice looking' Python code.

Reply:
There are some flags for the Jinaj2 environment. However, none seem to dedent correctly. I think creating and debugging an extension is a bit to much atm.

Reply:
Oh sorry, I didn't look properly. Confusing without the indents, indeed, but not worth worrying about it now, I would say.

Reply:
I've opted to add the `keep_trailing_newline` as I think it is useful: IIRC files without a trailing newlines are not valid files under POSIX.